Add function to retrieve a writable reference to particle locations

gassmoeller opened the pull request:
Add a function to retrieve a writable reference to the particle location and particle reference location. This is useful if one wants to perform some operations on the the same memory without copying the location into a temporary Point. For performance critical algorithms this can make a difference, and also the code just looks shorter and nicer. We already have a similar function for the particle properties.

All the changes are essentially just copies from the existing functions removing the `const` qualifier. Let me know if you prefer to have tests as well.

blaisb reviewed (approved):
No comments on my side, this seems like a healthy change :)

bangerth reviewed (commented):
Do I understand correctly that you want to be able to replace something like
```
  x.set_location (moved_point_coordinates(x.get_location()));
```
by
```
  move_this_point (x.get_location());
```
? Is copying 3 `double`s twice really a bottleneck in these cases?

I'm asking because `set_location()` has the advantage to at least in principle allow us to check for the validity of arguments. I don't know whether we do that right now -- presumably, we let users set any location, including ones where location and reference-location are inconsistent with each other, and then we require calling `particle_handler.sort_particles_into_cells()` or some such. If so, making particle locations write-able does not make it worse.

So perhaps the review should just be:
* Yes, add/adapt a test please
* Perhaps say in the documentation of the non-`const` version that this function allows writing into the returned point location, but that like with `set_location()`, one eventually has to call functions again that sort particles into the correct cells.

kronbichler commented:
I think this makes sense. I'm wondering if the interface would then be more consistent if the existing `const`-qualified functions would return `Point<dim>` by value instead of by reference. At the same time, I can imagine that there are some already existing codes that might explicitly take a `const Point<dim>&` variable, and that would stop working when done as suggested. So I tend towards keeping the suggested code.

gassmoeller commented:
I have added tests and replaced the usage of `get_location` within the library in the places where I think it is useful. I left step-19 unchanged as it is the simplest introduction to the particles, but did include it in step-68, which also shows some other optimizations for particles. I also removed the writable function for `get_reference_location` again, because I didnt find a good use for it in the library (and I also do not need it myself). I think the change is much more useful for `get_location`anyway.

To your questions Wolfgang:
> Do I understand correctly that you want to be able to replace something like
>  x.set_location (moved_point_coordinates(x.get_location()));
> by
>  move_this_point (x.get_location());?

Yes, I also want to allow things like:
```
x.get_location() += other_vector
```

I think this makes the code easier to read.

> Is copying 3 doubles twice really a bottleneck in these cases?

copying the 3 doubles is unfortunately not the only cost of the `set_location` function. It also has to retrieve the property pool handle of the particle and then ask the property pool for the reference. The propertypool then checks that the handle is non zero (even in optimized mode). All in all I save about 10% of the compute time of [this function](https://github.com/geodynamics/aspect/blob/b6d5e48fb6c29a1056c415211f60e0c332e13c9f/source/particle/integrator/rk_2.cc#L52) if I remove the two calls to `set_location`.

> I'm asking because set_location() has the advantage to at least in principle allow us to check for the validity of arguments. I don't know whether we do that right now

I agree that error checking would be an advantage, but as you say we do not do it right now, and I dont think it would be a good idea performance-wise either. I tend to think about the particle locations and properties like the entries in a solution vector and we dont do individual error checking there either.

> So perhaps the review should just be: Add a test. Add to the documentation.

Done.

@kronbichler: I agree, but I think the change would cause too many incompatibilities to be worth it.

Let me know if you think the current version is ok.

bangerth reviewed (approved):
Thanks. I will add that this mirrors `cell->vertex(i)`, which also returns a writable reference.

marcfehling commented:
/rebuild
